File: demo/benchmarker/compute.swift
import Foundation
import CryptoKit

/// Compute SHA256 hash of data
public func hashBytes(
    data: UnsafeRawPointer,
    len: Int,
    output: UnsafeMutableRawPointer
) {
    let buffer = UnsafeRawBufferPointer(start: data, count: len)
    let digest = SHA256.hash(data: buffer)
    digest.withUnsafeBytes { bytes in
        output.copyMemory(from: bytes.baseAddress!, byteCount: 32)
    }
}

/// Verify if data matches expected hash
public func verifyHash(
    data: UnsafeRawPointer,
    len: Int,
    expected: UnsafeRawPointer
) -> Bool {
    let buffer = UnsafeRawBufferPointer(start: data, count: len)
    let digest = SHA256.hash(data: buffer)
    
    let expectedBuffer = UnsafeRawBufferPointer(start: expected, count: 32)
    return digest.withUnsafeBytes { actual in
        return actual.elementsEqual(expectedBuffer)
    }
}

/// Compare two hashes for equality
public func compareHashes(
    hash1: UnsafeRawPointer,
    hash2: UnsafeRawPointer
) -> Bool {
    let buffer1 = UnsafeRawBufferPointer(start: hash1, count: 32)
    let buffer2 = UnsafeRawBufferPointer(start: hash2, count: 32)
    return buffer1.elementsEqual(buffer2)
}

File: demo/benchmarker/compute.zig
const std = @import("std");
const sha256 = std.crypto.hash.sha2.Sha256;

/// Compute SHA256 hash of data
export fn hashBytes(data: [*]const u8, len: usize, output: *[32]u8) void {
    sha256.hash(data[0..len], output, .{});
}

/// Verify if data matches expected hash
export fn verifyHash(data: [*]const u8, len: usize, expected: *const [32]u8) bool {
    var actual: [32]u8 = undefined;
    sha256.hash(data[0..len], &actual, .{});
    return std.mem.eql(u8, &actual, expected);
}

/// Compare two hashes for equality
export fn compareHashes(hash1: *const [32]u8, hash2: *const [32]u8) bool {
    return std.mem.eql(u8, hash1, hash2);
}
